File: backend/app/schemas/insights.py
from pydantic import BaseModel
from typing import Optional


class MonthComparison(BaseModel):
    current_month: str
    previous_month: Optional[str] = None
    current_expenses: float
    previous_expenses: Optional[float] = None
    change_amount: Optional[float] = None
    change_percent: Optional[float] = None


class CategoryChange(BaseModel):
    category: str
    current_amount: float
    previous_amount: float
    change_amount: float


class RecurringExpense(BaseModel):
    merchant: str
    average_amount: float
    occurrence_count: int
    frequency: str  # "monthly" (only type we detect for now)


class UnusualTransaction(BaseModel):
    id: int
    description: str
    merchant: Optional[str]
    amount: float
    transaction_date: str
    reason: str


class Insight(BaseModel):
    type: str  # "spending_increase", "spending_decrease", "recurring", "unusual", "saving_opportunity"
    message: str


class InsightsResponse(BaseModel):
    insights: list[Insight]
    month_comparison: Optional[MonthComparison] = None
    category_changes: list[CategoryChange] = []
    recurring_expenses: list[RecurringExpense] = []
    unusual_transactions: list[UnusualTransaction] = []

File: frontend/src/api/insights.ts
import api from "./client";

export interface MonthComparison {
  current_month: string;
  previous_month: string | null;
  current_expenses: number;
  previous_expenses: number | null;
  change_amount: number | null;
  change_percent: number | null;
}

export interface CategoryChange {
  category: string;
  current_amount: number;
  previous_amount: number;
  change_amount: number;
}

export interface RecurringExpense {
  merchant: string;
  average_amount: number;
  occurrence_count: number;
  frequency: string;
}

export interface UnusualTransaction {
  id: number;
  description: string;
  merchant: string | null;
  amount: number;
  transaction_date: string;
  reason: string;
}

export interface Insight {
  type: string;
  message: string;
}

export interface InsightsResponse {
  insights: Insight[];
  month_comparison: MonthComparison | null;
  category_changes: CategoryChange[];
  recurring_expenses: RecurringExpense[];
  unusual_transactions: UnusualTransaction[];
}

export async function getInsights(): Promise<InsightsResponse> {
  const response = await api.get("/insights");
  return response.data;
}